wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/validator/DevfileIntegrityValidator.java

```diff
@@ -149,4 +200,75 @@ private void validateProjects(Devfile devfile) throws DevfileFormatException {
       }
     }
   }
+
+  private List<HasMetadata> validateSelector(Tool tool, FileContentProvider contentProvider)
+      throws ValidationException, InfrastructureException, IOException, DevfileException {
+
+    List<HasMetadata> content = getReferencedKubernetesList(tool, contentProvider);
+
+    Map<String, String> selector = tool.getSelector();
+    if (selector != null) {
+      content = SelectorFilter.filter(content, selector);
+    }
+
+    if (content.isEmpty()) {
+      throw new DevfileException(
+          format(
+              "The selector of the tool %s filters out all objects from the list.",
+              tool.getName()));
+    }
+
+    return content;
+  }
+
+  private void validateEntrypointSelector(Tool tool, List<HasMetadata> filteredObjects)
+      throws DevfileFormatException {
+
+    if (tool.getEntrypoints() == null || tool.getEntrypoints().isEmpty()) {
+      return;
+    }
+
+    for (Entrypoint ep : tool.getEntrypoints()) {
+      ContainerSearch search =
+          new ContainerSearch(
+              ep.getDeployment(),
+              ep.getPod(),
+              ep.getContainer(),
+              ep.getPodSelector(),
+              ep.getDeploymentSelector());
+
+      List<Container> cs = search.search(filteredObjects);
+
+      if (cs.isEmpty()) {
+        throw new DevfileFormatException(
+            format(
+                "Tool %s contains an entry point that doesn't match any container: %s",
+                tool.getName(), toYAML(ep)));
```

Comment:
Please check that yaml inlined in error message looks good enough.

Reply:
Yes it does and now we don't even need to care about the formatting anymore. I handed it over to the `Serialization` class provided by k8s client ;)